precompiles/xcm/src/lib.rs

```diff
@@ -365,16 +433,567 @@ where
+        let currency_address = input.read::<Address>()?;
+        let amount_of_tokens = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting amount_of_tokens, maybe value too large"))?;
+        let fee = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("can't convert fee"))?;
+
+        let destination = input.read::<MultiLocation>()?;
+        let weight = input.read::<WeightV2>()?;
+
+        let asset_id = Runtime::address_to_asset_id(currency_address.into())
+            .ok_or(revert("Failed to resolve fee asset id from address"))?;
+        let dest_weight_limit = if weight.is_max() {
+            WeightLimit::Unlimited
+        } else {
+            WeightLimit::Limited(weight.get_weight())
+        };
+
+        log::trace!(target: "xcm-precompile::transfer_with_fee", "Raw arguments: currency_address: {:?}, amount_of_tokens: {:?}, destination: {:?}, \
+        weight: {:?}, calculated asset_id: {:?}",
+        currency_address, amount_of_tokens, destination, weight, asset_id);
+
+        let call = orml_xtokens::Call::<Runtime>::transfer_with_fee {
+            currency_id: asset_id.into(),
+            amount: amount_of_tokens,
+            fee,
+            dest: Box::new(VersionedMultiLocation::V3(destination)),
+            dest_weight_limit,
+        };
+
+        let origin = Some(Runtime::AddressMapping::into_account_id(
+            handle.context().caller,
+        ))
+        .into();
+
+        // Dispatch a call.
+        RuntimeHelper::<Runtime>::try_dispatch(handle, origin, call)?;
+
+        Ok(succeed(EvmDataWriter::new().write(true).build()))
+    }
+
+    fn transfer_multiasset(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+        let mut input = handle.read_input()?;
+        input.expect_arguments(5)?;
+
+        // Read call arguments
+        let asset_location = input.read::<MultiLocation>()?;
+        let amount_of_tokens = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting amount_of_tokens, maybe value too large"))?;
+        let destination = input.read::<MultiLocation>()?;
+        let weight = input.read::<WeightV2>()?;
+
+        let dest_weight_limit = if weight.is_max() {
+            WeightLimit::Unlimited
+        } else {
+            WeightLimit::Limited(weight.get_weight())
+        };
+
+        log::trace!(target: "xcm-precompile::transfer_multiasset", "Raw arguments: asset_location: {:?}, amount_of_tokens: {:?}, destination: {:?}, \
+        weight: {:?}",
+        asset_location, amount_of_tokens, destination, weight);
+
+        let call = orml_xtokens::Call::<Runtime>::transfer_multiasset {
+            asset: Box::new(VersionedMultiAsset::V3(MultiAsset {
+                id: AssetId::Concrete(asset_location),
+                fun: Fungibility::Fungible(amount_of_tokens),
+            })),
```

Comment:
To create a `MultiAsset`, you can use `(asset_location, amount_of_tokens).into()`. Same to all `MultiAsset` usages below.